Feat: Implement `MemoryStream` abstraction for user space memory access

caiyih opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Memory-streaming API with cursor reads, seeking, optional buffer retention, automatic sync/unmap on drop, and cross-MMU stream creation.
  * Native backend: cross-window support to enable cross-MMU mappings.

* **Breaking Changes**
  * Range operation APIs now take the second range by value.
  * New MMU/Paging error variant and new cross-mapping MMU methods.

* **Chores**
  * Added streaming crate to the workspace and updated manifests/dependencies.

* **Tests**
  * Expanded tests for mutability, cross-MMU behavior, and streaming.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Cai1Hsu committed:
refactor(address): remove unnecessary references in AddressRange methods

- Update AddressRange methods to accept AddressRange<T> instead of &AddressRange<T>
- This change simplifies the API and improves code readability
- Affected methods: contains_range, contained_by, intersects, intersection, union
- Updated related test cases and utilities

Cai1Hsu committed:
build(mmu-abstractions): remove unused downcast-rs dependency

- Remove commented-out dependency on downcast-rs
- This change simplifies the Cargo.toml file and removes unnecessary code

Cai1Hsu committed:
refactor(test-utilities): improve buffer mapping handling for TestMMU

- Add mutable flag to MappedMemory to track buffer mutability
- Update map_buffer_internal and map_buffer_internal_mut to use the mutable flag
- Modify unmap_buffer to handle mutable mappings differently
- Ensure proper synchronization of mutable memory to physical memory

Cai1Hsu committed:
feat(libraries): add stream library for memory operations

- Implement MemoryStream struct for reading and writing memory
- Add support for reading slices and unsized types
- Include cursor manipulation and synchronization methods
- Update Cargo.toml to include new library

Cai1Hsu committed:
feat(mmu): implement cross-MMU mapping functionality

- Add map_cross_internal and map_cross_mut_internal methods to IMMU trait
- Implement cross-MMU mapping support in PageTableNative
- Update TestMMU for testing cross-MMU mapping
- Add downcast-rs dependency for dynamic casting

Cai1Hsu committed:
Merge branch 'master' into feat/streamed-read-memory

Cai1Hsu committed:
fix(mmu-native): correct condition for removing cross mapping window

- Update the condition check in CrossMappingAllocator::remove method
- Change from negation to positive check for vaddr range inclusion
- This fixes the logic error in removing the correct mapping window

Cai1Hsu committed:
fix various issues

Cai1Hsu committed:
fix lint issues

Cai1Hsu committed:
fix(mmu-native): correct slice creation in query_physical_page

Cai1Hsu committed:
refactor(stream): add window mapping and alignment check for stream access

- Add logic to check for overlap with existing window
- Implement prevention of mapping leaks by storing mapped window information
- Add alignment check for mapped buffer

Cai1Hsu committed:
fix(stream): handle borrowed buffer keep on window remap

Cai1Hsu committed:
feat(stream): add cross MMU support and optimize memory handling

- Implement new_cross function for creating memory stream with cross MMU
- Add source and mmu_map_buffer helper functions
- Refactor existing code to support both single and cross MMU scenarios
- Optimize buffer handling and improve error checking

Cai1Hsu committed:
fix(stream): correct mapping size and handle empty streams for `Stream::window`

- Update window mapping logic to handle empty streams properly
- Adjust buffer reuse condition for full range checks
- Modify window initialization to use correct base and length
- Implement offset calculation for mapped buffer to ensure proper slice indexing

Cai1Hsu committed:
refactor(stream): replace MemoryStream with generic Stream type

- Remove specific MemoryStream struct
- Introduce generic Stream type using macro_rules
- Keep existing functionality and naming conventions

Cai1Hsu committed:
fix(stream): add alignment check for stream read and write operations

- Add alignment check for cursor position before performing read and write operations
- Use `self.cursor()` instead of `self.inner().cursor` to get the cursor position
- Improve error handling by returning `MMUError::MisalignedAddress` for misaligned addresses

Cai1Hsu committed:
refactor(stream): improve Stream read API and add new methods

- Rename `read_slice` to `read_slice_internal` and make it private
- Add public `read_slice` method that moves the cursor after reading
- Add public `pread_slice` method for reading without moving the cursor
- Modify `read` method to use `read_slice_internal` and move the cursor
- Add `pread` method for reading without moving the cursor
- Rename `read_unsized` to `read_unsized_internal` and make it private
- Add public `read_unsized_slice` method that moves the cursor after reading
- Add public `pread_unsized_slice` method for reading without moving the cursor

Cai1Hsu committed:
refactor(stream): split read_slice_internal into inspect_slice_internal and read_slice_internal

- Rename `read_slice_internal` to `inspect_slice_internal` for better naming convention
- Split `inspect_slice_internal` into two separate functions:
  - `inspect_slice_internal`: returns a raw pointer and length
  - `read_slice_internal`: uses `inspect_slice_internal` and returns a slice reference
- Update `inspect_slice_internal` to return a Result containing a NonNull pointer and length
- Modify `read_slice_internal` to use `inspect_slice_internal` and create a slice reference
- Update code to use new function names and return types

Cai1Hsu committed:
test(stream): add comprehensive unit tests for MemoryStream

- Implement tests for stream creation and cursor operations
- Add tests for reading different data types and slice sizes
- Implement tests for unsized reads and cross-MMU functionality
- Add tests for edge cases like misaligned addresses and read-only memory
- Implement tests for buffer keep functionality and window reuse
- Add tests for large reads and seek boundaries
- Implement tests for consecutive reads and mixed read operations

Cai1Hsu committed:
feat(stream): add write and pwrite methods for MemoryStreamMut

Cai1Hsu committed:
test(stream): simplify assertions in read_slice tests for clarity

Cai1Hsu committed:
feat(mmu): implement unmapping for cross-mapped buffers

- Add unmap_cross method to IMMU trait
- Implement unmap_cross in PageTable for removing cross-mappings
- Add support for unmap_cross in TestMMU for testing

Cai1Hsu committed:
refactor(stream): optimize buffer mapping for write access

- Introduce mmu_map_buffer_mut function for mutable buffer mapping
- Implement unmap function to centralize buffer unmapping logic
- Update sync and unmap_current functions to use the new unmap method
- Modify check_full_range to select appropriate mapping function based on access type

Cai1Hsu committed:
test(stream): add comprehensive tests for MemoryStreamMut

- Implement numerous test cases for MemoryStreamMut functionality
- Cover basic creation, write operations, slice writes, and mixed operations
- Test edge cases like empty slices and misaligned addresses
- Verify behavior with read-only memory and invalid addresses
- Check buffer keep functionality and cross-MMU writes
- Ensure proper cleanup and data consistency after operations

Cai1Hsu committed:
test(stream): remove unnecessary sync calls in read_slice tests

Cai1Hsu committed:
refactor(stream): simplify MMU stream initialization

- Remove unnecessary Option for cross MMU parameter
- Simplify new_cross method implementation
- Update MmuComposition usage in stream creation

Cai1Hsu committed:
feat(stream): add mutable stream creation methods to IMMUStreamExt trait

Cai1Hsu committed:
refactor(tests): replace helper functions with direct calls to create_stream_mut

Cai1Hsu committed:
fix(mmu): prevent map leak in `map_buffer_internal`

Cai1Hsu committed:
fix(mmu-native): use correct variable in query_virtual call

- Replace `vaddr` with `checking` in query_virtual call within the page table implementation
- This change ensures that the correct virtual address is used for querying page table entries

Cai1Hsu commented on libraries/mmu-native/src/pt.rs:
```diff
@@ -270,9 +348,157 @@ impl<Arch: IPageTableArchAttribute, PTE: IArchPageTableEntry> IMMU for PageTable
+        loop {
+            let (phy, permission, sz) = source.query_virtual(checking).map_err(|e| e.into())?;
+
+            ensure_permission(vaddr, permission, false)?;
+            phys.push((phy, sz));
+
+            let page_offset = vaddr.as_usize() % sz.as_usize();
+            if slice_offset.is_none() {
+                slice_offset = Some(page_offset);
+            }
+
+            let sz = sz.as_usize();
+
+            // align to page size
+            checking -= page_offset;
+
+            if checking + sz >= end {
+                let vaddr = *window.deref();
+                window.cancel(); // prevent drop
+
+                drop(cross);
+
+                let mut offset = 0;
+                for (phy, sz) in phys {
+                    self.map_single(vaddr + offset, phy, sz, PERMISSION)
+                        .unwrap();
+                    offset += sz.as_usize();
+                }
+
+                return Ok(unsafe {
+                    core::slice::from_raw_parts(
+                        vaddr.as_ptr::<u8>().add(slice_offset.unwrap()),
+                        len,
+                    )
+                });
+            }
+
+            checking += sz;
+        }
+    }
```
Using `checking` for query_virtual is correct, i've fixed the map_cross_mut_internal method and pushed the fix. `ensure_permission` only takes the queried page flags and compares with the given mutability info, the provided vaddr is used for the error message hint. vaddr would be clearer when debugging the error.

Cai1Hsu commented on libraries/stream/src/lib.rs:
```diff
@@ -0,0 +1,1685 @@
+        cursor: VirtualAddress,
+        keep_buffer: bool,
+    ) -> MemoryStream<'a> {
+        MemoryStream::new_cross(self, src, cursor, keep_buffer)
```
it was a typo, the second parameter should be the source MMU, my bad

Cai1Hsu replied:
it was `Option`, the doc needs to be updated

Cai1Hsu replied:
it was `Option`, the doc needs to be updated

Cai1Hsu committed:
fix typo

Cai1Hsu committed:
perf(stream): optimize memory mapping performance

- Add comment explaining trade-off between security and performance in memory mapping
- Consider using checking functions' access permission to prevent unnecessary remapping
- Note potential impact on compiler optimizations and dead code elimination

Cai1Hsu commented on libraries/stream/src/lib.rs:
```diff
@@ -0,0 +1,1685 @@
+                // Move to a different position, the buffer should be preserved
+                stream.seek(Whence::Set(base + 8));
+                // Since the buffer overlaps, this may fail, which is expected
+                let result2 = stream.read_slice::<i32>(2);
+
+                // If no error occurs, validate the data correctness
+                let slice = result2.unwrap();
+                assert_eq!(slice.len(), 2);
+            }
```
This test was intended to test that we keep the stream's reference, and test if the reference is still accessible or not when the stream's cursor goes out of the current page. But such an internal state can NOT be asserted outside, and it's unnecessary to do so.

The `MMUError::Borrowed` condition is not reproduced in such a simple way. This can only be reproduced when the buffer is already mapped and the mapped buffer is smaller than the buffer size specified in this call. But, in our case, whenever possible, even if in a read method, we map the buffer with write permission. I've added a comment in the latest commit for this. The only way to reproduce it is that we first map the buffer outside, with read-only permission, then create a new MemoryStreamMut and call write methods on it. The buffer is usually mapped by whole page, and I was considering adding write permission in buffer mappings when possible, and trying to reuse mapped buffers. This situation wouldn't exist if that was implemented. And we expect the MMU to be locked if we want to map a buffer, and unmap all existing buffers when we unlock the MMU. So this case is very rare even in the current situation. We can add such a test, but I always feel it's unnecessary.

Cai1Hsu committed:
docs(stream): enhance documentation for memory stream library and interfaces

Cai1Hsu committed:
fix(stream): optimize buffer unmapping logic in sync method

Cai1Hsu committed:
fix(stream): ensure current buffer is unmapped

Cai1Hsu committed:
docs(stream): clarify documentation for reading variable-length sequences from the stream
